File: packages/adapters/src/venues/bitmart.ts
import {
  type AssetClass,
  canonicalBase,
  classifyNonCrypto,
  type FundingEvent,
  type FundingSnapshot,
  inferIntervalHours,
  type MarketRef,
} from "@ai-rates/core";
import type { HttpClient } from "../http";
import { marketRef, mul, num } from "../parse";
import type { SnapshotBatch, VenueAdapter } from "../types";

const VENUE_ID = "bitmart";
const BASE_URL = "https://api-cloud-v2.bitmart.com/contract/public";
const OK = 1000;
/** `product_type` 1 is a perpetual (2 would be a dated future); every row was 1 on 2026-09-14. */
const PERPETUAL = 1;
/** `funding-rate-history` caps `limit` at 100 (1000 answers 100) and has no way to page further. */
const HISTORY_LIMIT = 100;

export interface BitmartEnvelope<T> {
  code: number;
  message: string;
  data: T;
}

export interface BitmartTradfiInfo {
  /**
   * The market a tradfi contract follows: US_MARKET, HK_STOCK, FOREX, INDEX_{JP,HK,KR,UK,AU,TW,DE},
   * METAL_LME, COMMODITY_{CME,ICE} and PRE_LIST on 2026-09-14.
   */
  market_group?: string;
}

export interface BitmartContract {
  symbol: string;
  product_type: number;
  base_currency: string;
  /** Settlement currency: USDT, USDC, or USD on the coin-margined inverse contracts. */
  quote_currency: string;
  index_price: string;
  /** Base units per contract. */
  contract_size: string;
  /** Open interest in contracts. */
  open_interest: string;
  /** Quote currency. */
  turnover_24h: string;
  /** Estimate for the settlement at `funding_time`. */
  expected_funding_rate: string;
  /** Epoch ms of the next settlement. */
  funding_time: number;
  funding_interval_hours: number;
  max_leverage?: string;
  status: string;
  /** Epoch SECONDS; 0 when none is scheduled. */
  delist_time: number;
  /** Null on crypto contracts. */
  tradfi_info?: BitmartTradfiInfo | null;
}

export interface BitmartFundingHistoryItem {
  symbol: string;
  funding_rate: string;
  /** Epoch ms of the settlement. */
  funding_time: string;
}

function unwrap<T>(json: BitmartEnvelope<T>, what: string): T {
  if (json?.code !== OK || json.data === null || typeof json.data !== "object") {
    throw new Error(`bitmart ${what}: ${json?.code} ${json?.message ?? ""}`);
  }
  return json.data;
}

/**
 * The class BitMart declares for a contract, from `tradfi_info.market_group` on /details.
 *
 * `tradfi_info` is null on every crypto contract and set on every tradfi one. Live 2026-09-14, of
 * 228 tradable contracts: none on 151; US_MARKET 49 (shares, ETFs, and SPX500, NAS100, US30 and
 * US2000, which `marketRef` refines to index); HK_STOCK 14; FOREX 5 (EUR, JPY, GBP, TRY, BRL);
 * INDEX_JP 2, INDEX_KR 2, INDEX_HK 1, INDEX_TW 1 (JPN225 and TW88 stay index; KIOXIA, SKHYNIX, HPSP
 * and TRAHK are single names and refine to equity); METAL_LME 3 (XNI, XCU, XAL). That comes out as
 * crypto 151, equity 63, index 6, fx 5, commodity 3.
 *
 * BitMart attaches no tradfi_info to XAUUSDT, XAGUSDT, PAXGUSDT, XAUTUSDT or CLUSDT, so gold, silver
 * and crude are crypto here, by the venue's own declaration. A group this does not know is still
 * tradfi.
 */
export function bitmartAssetClass(
  tradfi: BitmartTradfiInfo | null | undefined,
  base: string,
): AssetClass {
  if (!tradfi) return "crypto";
  const group = tradfi.market_group ?? "";
  if (group === "US_MARKET" || group === "HK_STOCK") return "equity";
  if (group === "FOREX") return "fx";
  if (group.startsWith("INDEX_")) return "index";
  if (group.startsWith("METAL_") || group.startsWith("COMMODITY_")) return "commodity";
  return classifyNonCrypto(canonicalBase(base));
}

/**
 * Live USDT- and USDC-margined perpetuals.
 *
 * Live 2026-09-14, 1,215 rows: 856 Delisted and 359 Trading. Of the Trading ones, 127 carry a
 * `delist_time` in the past (all 2026-07-25) and are dead: zero turnover on every one, and empty
 * books on THETAUSDT, ARBUSDT, HK50USDT and TENCENTUSDT. Four more are coin-margined inverse
 * contracts quoted in USD (BTCUSD, ETHUSD, XRPUSD, SOLUSD, 10-100 USD a contract). That leaves 228:
 * 227 USDT and BTCUSDC. A future `delist_time` is still trading and stays in.
 */
export function isBitmartTradable(contract: BitmartContract, now: number): boolean {
  return (
    contract.status === "Trading" &&
    contract.product_type === PERPETUAL &&
    (contract.quote_currency === "USDT" || contract.quote_currency === "USDC") &&
    !(contract.delist_time > 0 && contract.delist_time * 1000 <= now)
  );
}

/**
 * Every tradable contract from the one bulk /details response.
 *
 * The rate is `expected_funding_rate`, the estimate for `funding_time`. The row's `funding_rate` is
 * not used: it is neither that estimate nor the last settlement. Measured 2026-09-14 against the
 * per-symbol /funding-rate, whose `expected_rate` matches `expected_funding_rate` (SOL 0.0000725,
 * DOGE 0.0000631, AXS 0.00005) and whose `rate_value` matches the newest /funding-rate-history row
 * (SOL 0.0001, DOGE 0.0004), `funding_rate` read -0.0000111 for SOL and -0.0000034 for DOGE.
 *
 * /details carries no mark price and there is no bulk endpoint that does (only a per-symbol mark
 * kline), so `markPrice` is null and open interest is priced at the index.
 */
export function parseBitmartSnapshots(
  json: BitmartEnvelope<{ symbols: BitmartContract[] }>,
  now: number,
): SnapshotBatch {
  const { symbols } = unwrap(json, "details");
  if (!Array.isArray(symbols)) throw new Error("bitmart details: expected a symbol list");

  const snapshots: FundingSnapshot[] = [];
  for (const contract of symbols) {
    if (!isBitmartTradable(contract, now)) continue;
    const rate = num(contract.expected_funding_rate);
    const hours = num(contract.funding_interval_hours);
    if (rate === null || hours === null || hours <= 0) continue;

    const { base } = marketRef(VENUE_ID, contract.symbol);
    const indexPrice = num(contract.index_price);
    const next = num(contract.funding_time);
    snapshots.push({
      ...marketRef(VENUE_ID, contract.symbol, {
        quote: contract.quote_currency,
        assetClass: bitmartAssetClass(contract.tradfi_info, base),
      }),
      observedAt: now,
      rate,
      basisHours: hours,
      intervalHours: hours,
      nextFundingAt: next !== null && next > 0 ? next : null,
      kind: "predicted",
      markPrice: null,
      indexPrice,
      // Contracts x base units per contract x price. BTCUSDT: 2,056,722 x 0.001 BTC x 77,296 =
      // $159.0m. `open_interest_value` ($155.0m there) is not used: on 179 contracts it sits 0.68x
      // to 1.55x off contracts x size x price -- BCHUSDT implies $329 a coin at a $224 price -- which
      // reads as entry notional, not the value of the open positions now.
      openInterestUsd: mul(num(contract.open_interest), num(contract.contract_size), indexPrice),
      // Quote turnover: BTCUSDT's 1,544,041,296 is 20,047,458 contracts x 0.001 x ~77,250.
      volume24hUsd: num(contract.turnover_24h),
      maxLeverage: num(contract.max_leverage),
    });
  }
  return { snapshots, settled: [] };
}

/**
 * Settled funding inside [fromMs, toMs], oldest first, with the interval inferred from every row
 * fetched so a one-event window still gets a basis.
 */
export function parseBitmartFundingHistory(
  ref: MarketRef,
  items: readonly BitmartFundingHistoryItem[],
  fromMs: number,
  toMs: number,
  fallbackHours: number | null,
): FundingEvent[] {
  const rates = new Map<number, number>();
  for (const item of items) {
    const settledAt = num(item.funding_time);
    const rate = num(item.funding_rate);
    if (settledAt !== null && settledAt > 0 && rate !== null) rates.set(settledAt, rate);
  }
  const basisHours = inferIntervalHours([...rates.keys()]) ?? fallbackHours;
  if (basisHours === null) return [];

  return [...rates]
    .filter(([settledAt]) => settledAt >= fromMs && settledAt <= toMs)
    .sort((a, b) => a[0] - b[0])
    .map(([settledAt, rate]) => ({ ...ref, settledAt, rate, basisHours, markPrice: null }));
}

function refOf(snapshot: FundingSnapshot): MarketRef {
  const { venueId, venueSymbol, base, assetClass, quote, multiplier, dex } = snapshot;
  return { venueId, venueSymbol, base, assetClass, quote, multiplier, dex };
}

/** BitMart's USDT- and USDC-margined perpetuals: one request a cycle. */
export function createBitmartAdapter(): VenueAdapter {
  let known = new Map<string, { ref: MarketRef; intervalHours: number | null }>();

  return {
    venueId: VENUE_ID,
    // Public contract endpoints answer `x-bm-ratelimit-limit: 12`, `x-bm-ratelimit-reset: 2`,
    // mode IP: 12 requests per 2 seconds. One request each 200ms is 10 per 2 seconds.
    minIntervalMs: 200,

    async fetchSnapshots(client: HttpClient, now: number) {
      const batch = parseBitmartSnapshots(
        await client.getJson<BitmartEnvelope<{ symbols: BitmartContract[] }>>(
          `${BASE_URL}/details`,
        ),
        now,
      );
      known = new Map(
        batch.snapshots.map((s) => [
          s.venueSymbol,
          { ref: refOf(s), intervalHours: s.intervalHours },
        ]),
      );
      return batch;
    },

    /**
     * The newest 100 settlements, cut to the window. `start_time` and `end_time` are ignored (a
     * 2026-06 window answered with the latest five) and `limit` stops at 100, so this reaches back
     * 100 intervals -- 33 days at 8h, four at 1h -- and a backfill older than that gets nothing,
     * which the collector records as the venue's limit. Coverage can also be stale: ESPORTSUSDT,
     * settling hourly on 2026-09-14, had no history row after 2026-07-30.
     */
    async fetchFundingHistory(client, venueSymbol, fromMs, toMs) {
      const json = await client.getJson<BitmartEnvelope<{ list: BitmartFundingHistoryItem[] }>>(
        `${BASE_URL}/funding-rate-history?symbol=${encodeURIComponent(venueSymbol)}&limit=${HISTORY_LIMIT}`,
      );
      const list = unwrap(json, "funding history").list ?? [];
      const market = known.get(venueSymbol);
      return parseBitmartFundingHistory(
        market?.ref ?? marketRef(VENUE_ID, venueSymbol),
        list,
        fromMs,
        toMs,
        market?.intervalHours ?? null,
      );
    },
  };
}

export const bitmartAdapter = createBitmartAdapter();

File: apps/collector-go/internal/adapters/bitmart/bitmart.go
// Package bitmart parses BitMart's USDT- and USDC-margined perpetuals.
//
// Ported from packages/adapters/src/venues/bitmart.ts. Two venue facts run through everything here:
//
//   - THERE IS NO MARK PRICE IN ANY BULK CALL. /details does not carry one and no other bulk endpoint
//     does either (only a per-symbol mark kline), so every snapshot's MarkPrice is nil and open
//     interest is priced at the index instead. A market with no mark is a ROUTINE state on this
//     venue, not an error, which is why migration 020 made the identity gate fall back to the index
//     price for it. Absent must stay nil: decoding it as 0 would claim a free market.
//   - THE RATE IS `expected_funding_rate`, NOT `funding_rate`. The row's `funding_rate` is neither
//     the estimate for the next settlement nor the last one paid; see ParseSnapshots.
package bitmart

import (
	"fmt"
	"sort"
	"strings"

	"github.com/belyjelli/ai-rates/collector/internal/adapters"
	"github.com/belyjelli/ai-rates/collector/internal/core"
)

const VenueID = "bitmart"

// okCode is the envelope code BitMart answers a successful call with.
const okCode = 1000

// perpetual is `product_type` 1 (2 would be a dated future); every row was 1 on 2026-09-14.
const perpetual = 1

// Envelope is BitMart's uniform response wrapper.
//
// Data is a POINTER so that `"data": null` on an error envelope stays absent instead of decoding to
// a zero-valued payload — the Go equivalent of the TypeScript `data === null` guard, and the same
// absent-is-not-empty rule the numerics follow.
type Envelope[T any] struct {
	Code    int    `json:"code"`
	Message string `json:"message"`
	Data    *T     `json:"data"`
}

func (e Envelope[T]) unwrap(what string) (*T, error) {
	if e.Code != okCode || e.Data == nil {
		return nil, fmt.Errorf("bitmart %s: %d %s", what, e.Code, e.Message)
	}
	return e.Data, nil
}

// TradfiInfo is the block BitMart attaches to a contract that tracks something off-chain.
type TradfiInfo struct {
	// MarketGroup is the market a tradfi contract follows: US_MARKET, HK_STOCK, FOREX,
	// INDEX_{JP,HK,KR,UK,AU,TW,DE}, METAL_LME, COMMODITY_{CME,ICE} and PRE_LIST on 2026-09-14.
	MarketGroup string `json:"market_group"`
}

type Contract struct {
	Symbol       string `json:"symbol"`
	ProductType  int    `json:"product_type"`
	BaseCurrency string `json:"base_currency"`
	// QuoteCurrency is the settlement currency: USDT, USDC, or USD on the coin-margined inverse
	// contracts.
	QuoteCurrency string       `json:"quote_currency"`
	IndexPrice    adapters.Num `json:"index_price"`
	// ContractSize is base units per contract.
	ContractSize adapters.Num `json:"contract_size"`
	// OpenInterest is open interest in CONTRACTS.
	OpenInterest adapters.Num `json:"open_interest"`
	// Turnover24h is quote currency.
	Turnover24h adapters.Num `json:"turnover_24h"`
	// ExpectedFundingRate is the estimate for the settlement at FundingTime.
	ExpectedFundingRate adapters.Num `json:"expected_funding_rate"`
	// FundingTime is epoch ms of the next settlement.
	FundingTime          adapters.Num `json:"funding_time"`
	FundingIntervalHours adapters.Num `json:"funding_interval_hours"`
	MaxLeverage          adapters.Num `json:"max_leverage"`
	Status               string       `json:"status"`
	// DelistTime is epoch SECONDS; 0 when none is scheduled.
	DelistTime int64 `json:"delist_time"`
	// TradfiInfo is null on crypto contracts.
	TradfiInfo *TradfiInfo `json:"tradfi_info"`
}

// Details is the payload of /details.
type Details struct {
	Symbols []Contract `json:"symbols"`
}

type FundingHistoryItem struct {
	Symbol      string       `json:"symbol"`
	FundingRate adapters.Num `json:"funding_rate"`
	// FundingTime is epoch ms of the settlement, quoted as a string here and as a number on /details.
	FundingTime adapters.Num `json:"funding_time"`
}

// FundingHistory is the payload of /funding-rate-history.
type FundingHistory struct {
	List []FundingHistoryItem `json:"list"`
}

// AssetClassFor is the class BitMart declares for a contract, from `tradfi_info.market_group` on
// /details.
//
// `tradfi_info` is null on every crypto contract and set on every tradfi one. Live 2026-09-14, of
// 228 tradable contracts: none on 151; US_MARKET 49 (shares, ETFs, and SPX500, NAS100, US30 and
// US2000, which MarketRefFor refines to index); HK_STOCK 14; FOREX 5 (EUR, JPY, GBP, TRY, BRL);
// INDEX_JP 2, INDEX_KR 2, INDEX_HK 1, INDEX_TW 1 (JPN225 and TW88 stay index; KIOXIA, SKHYNIX, HPSP
// and TRAHK are single names and refine to equity); METAL_LME 3 (XNI, XCU, XAL). That comes out as
// crypto 151, equity 63, index 6, fx 5, commodity 3.
//
// BitMart attaches no tradfi_info to XAUUSDT, XAGUSDT, PAXGUSDT, XAUTUSDT or CLUSDT, so gold, silver
// and crude are crypto here, by the venue's own declaration. A group this does not know is still
// tradfi.
func AssetClassFor(tradfi *TradfiInfo, base string) core.AssetClass {
	if tradfi == nil {
		return core.ClassCrypto
	}
	group := tradfi.MarketGroup
	if group == "US_MARKET" || group == "HK_STOCK" {
		return core.ClassEquity
	}
	if group == "FOREX" {
		return core.ClassFX
	}
	if strings.HasPrefix(group, "INDEX_") {
		return core.ClassIndex
	}
	if strings.HasPrefix(group, "METAL_") || strings.HasPrefix(group, "COMMODITY_") {
		return core.ClassCommodity
	}
	return core.ClassifyNonCrypto(core.CanonicalBase(base))
}

// IsTradable reports whether a /details row is a live USDT- or USDC-margined perpetual.
//
// Live 2026-09-14, 1,215 rows: 856 Delisted and 359 Trading. Of the Trading ones, 127 carry a
// `delist_time` in the past (all 2026-07-25) and are dead: zero turnover on every one, and empty
// books on THETAUSDT, ARBUSDT, HK50USDT and TENCENTUSDT. Four more are coin-margined inverse
// contracts quoted in USD (BTCUSD, ETHUSD, XRPUSD, SOLUSD, 10-100 USD a contract). That leaves 228:
// 227 USDT and BTCUSDC. A future `delist_time` is still trading and stays in.
func IsTradable(contract Contract, now int64) bool {
	return contract.Status == "Trading" &&
		contract.ProductType == perpetual &&
		(contract.QuoteCurrency == "USDT" || contract.QuoteCurrency == "USDC") &&
		!(contract.DelistTime > 0 && contract.DelistTime*1000 <= now)
}

// ParseSnapshots normalises every tradable contract from the one bulk /details response.
//
// The rate is `expected_funding_rate`, the estimate for `funding_time`. The row's `funding_rate` is
// not used: it is neither that estimate nor the last settlement. Measured 2026-09-14 against the
// per-symbol /funding-rate, whose `expected_rate` matches `expected_funding_rate` (SOL 0.0000725,
// DOGE 0.0000631, AXS 0.00005) and whose `rate_value` matches the newest /funding-rate-history row
// (SOL 0.0001, DOGE 0.0004), `funding_rate` read -0.0000111 for SOL and -0.0000034 for DOGE.
//
// /details carries no mark price and there is no bulk endpoint that does (only a per-symbol mark
// kline), so MarkPrice is nil and open interest is priced at the index.
func ParseSnapshots(env Envelope[Details], now int64) (core.SnapshotBatch, error) {
	details, err := env.unwrap("details")
	if err != nil {
		return core.SnapshotBatch{}, err
	}
	// An absent `symbols` is not an empty book: `"symbols": []` decodes to a non-nil empty slice,
	// while a missing or null field leaves it nil, which is the Go reading of the TypeScript
	// Array.isArray guard.
	if details.Symbols == nil {
		return core.SnapshotBatch{}, fmt.Errorf("bitmart details: expected a symbol list")
	}

	snapshots := make([]core.FundingSnapshot, 0, len(details.Symbols))
	for _, contract := range details.Symbols {
		if !IsTradable(contract, now) {
			continue
		}
		rate := contract.ExpectedFundingRate
		hours := contract.FundingIntervalHours
		if !rate.OK || !hours.OK || hours.Val <= 0 {
			continue
		}

		base := adapters.MarketRefFor(VenueID, contract.Symbol, adapters.Overrides{}).Base
		class := AssetClassFor(contract.TradfiInfo, base)
		quote := contract.QuoteCurrency
		ref := adapters.MarketRefFor(VenueID, contract.Symbol, adapters.Overrides{
			Quote:      &quote,
			HasQuote:   true,
			AssetClass: &class,
		})

		indexPrice := contract.IndexPrice.Ptr()
		interval := hours.Val
		snapshots = append(snapshots, core.FundingSnapshot{
			MarketRef:     ref,
			ObservedAt:    now,
			Rate:          rate.Val,
			BasisHours:    hours.Val,
			IntervalHours: &interval,
			NextFundingAt: contract.FundingTime.PositiveMs(),
			Kind:          core.KindPredicted,
			// Nil, never zero: BitMart publishes no mark in any bulk call, so "no mark" is a routine
			// state here and the identity gate falls back to the index price (migration 020).
			MarkPrice:  nil,
			IndexPrice: indexPrice,
			// Contracts x base units per contract x price. BTCUSDT: 2,056,722 x 0.001 BTC x 77,296 =
			// $159.0m. `open_interest_value` ($155.0m there) is not used: on 179 contracts it sits 0.68x
			// to 1.55x off contracts x size x price -- BCHUSDT implies $329 a coin at a $224 price -- which
			// reads as entry notional, not the value of the open positions now.
			OpenInterestUSD: adapters.Mul(contract.OpenInterest.Ptr(), contract.ContractSize.Ptr(), indexPrice),
			// Quote turnover: BTCUSDT's 1,544,041,296 is 20,047,458 contracts x 0.001 x ~77,250.
			Volume24hUSD: contract.Turnover24h.Ptr(),
			MaxLeverage:  contract.MaxLeverage.Ptr(),
		})
	}
	return core.SnapshotBatch{Snapshots: snapshots, Settled: []core.FundingEvent{}}, nil
}

// ParseFundingHistory returns settled funding inside [fromMs, toMs], oldest first, with the interval
// inferred from every row fetched so a one-event window still gets a basis.
//
// fallbackHours is nil where nothing has ever reported an interval for the market; an event with no
// basis is dropped rather than given an invented one.
func ParseFundingHistory(
	ref core.MarketRef,
	items []FundingHistoryItem,
	fromMs, toMs int64,
	fallbackHours *float64,
) []core.FundingEvent {
	// One rate per settlement timestamp, keeping the venue's own ordering: the first occurrence keeps
	// its position and the last wins on value, exactly as the TypeScript Map does. A bare Go map would
	// reshuffle the inference input per run.
	rates := make(map[int64]float64, len(items))
	order := make([]int64, 0, len(items))
	for _, item := range items {
		settledAt := item.FundingTime.PositiveMs()
		if settledAt == nil || !item.FundingRate.OK {
			continue
		}
		if _, seen := rates[*settledAt]; !seen {
			order = append(order, *settledAt)
		}
		rates[*settledAt] = item.FundingRate.Val
	}

	basisHours := fallbackHours
	if inferred := core.InferIntervalHours(order); inferred != nil {
		basisHours = inferred
	}
	if basisHours == nil {
		return []core.FundingEvent{}
	}

	events := make([]core.FundingEvent, 0, len(order))
	for _, settledAt := range order {
		if settledAt < fromMs || settledAt > toMs {
			continue
		}
		events = append(events, core.FundingEvent{
			MarketRef:  ref,
			SettledAt:  settledAt,
			Rate:       rates[settledAt],
			BasisHours: *basisHours,
			MarkPrice:  nil,
		})
	}
	sort.SliceStable(events, func(i, j int) bool { return events[i].SettledAt < events[j].SettledAt })
	return events
}
